### PR TITLE
[SYCL] Exit early while trying to enqueue blocked tasks

### PR DESCRIPTION
**Pull request by DenisBakhvalov**

### CODE REVIEW

File: sycl/unittests/scheduler/LeafLimit.cpp [hunk 1]
```diff
@@ -18,40 +18,43 @@ using namespace cl::sycl;
...
 
   buffer<int, 1> Buf(range<1>(1));
   detail::Requirement MockReq = getMockRequirement(Buf);
-  MockCommand *MockDepCmd =
-      new MockCommand(detail::getSyclObjImpl(MQueue), MockReq);
+
+  MockDepCmd =
+      std::make_unique<MockCommand>(detail::getSyclObjImpl(MQueue), MockReq);
   detail::MemObjRecord *Rec =
       MS.getOrInsertMemObjRecord(detail::getSyclObjImpl(MQueue), &MockReq);
```
Comment (romanovvlad) [inline sycl/unittests/scheduler/LeafLimit.cpp:30]: Could you please commit these refactoring as a separate PR?

**Review by romanovvlad — COMMENTED** [review]
LGTM

File: sycl/unittests/scheduler/LeafLimit.cpp (hunk 1 above)
Comment (DenisBakhvalov) [reply sycl/unittests/scheduler/LeafLimit.cpp:30]: Those changes fix memory leaks in the `LeafLimit` test. They are exposed (as `Gtest` warnings) as a result of my changes to `MockCommand`. I initially planned to fix memory leaks later, but testing revealed new failures.

Comment (romanovvlad) [reply sycl/unittests/scheduler/LeafLimit.cpp:30]: Ok then.

File: sycl/unittests/scheduler/SchedulerTestUtils.hpp [hunk 2]
```diff
@@ -36,7 +47,14 @@ class MockCommand : public cl::sycl::detail::Command {
     return &MRequirement;
   };
 
-  cl_int enqueueImp() override { return MRetVal; }
+  cl_int enqueueImp() override { return MRetVal; };
```
Comment (s-kanaev) [inline sycl/unittests/scheduler/SchedulerTestUtils.hpp:50]:
```suggestion
  cl_int enqueueImp() override { return MRetVal; }
```

File: sycl/unittests/scheduler/SchedulerTestUtils.hpp [hunk 3]
```diff
@@ -36,7 +47,14 @@ class MockCommand : public cl::sycl::detail::Command {
...
+
+  MOCK_METHOD2(enqueue, bool(cl::sycl::detail::EnqueueResultT &,
+                             cl::sycl::detail::BlockingT));
+  bool enqueueOrigin(cl::sycl::detail::EnqueueResultT &EnqueueResult,
+                     cl::sycl::detail::BlockingT Blocking) {
+    return cl::sycl::detail::Command::enqueue(EnqueueResult, Blocking);
+  }
```
Comment (s-kanaev) [inline sycl/unittests/scheduler/SchedulerTestUtils.hpp:57]: As far as I understand the mock framework will catch the call to `MockCommand::enqueue` and call `MockCommand::enqueueOrigin` instead. The latter will call `Command::enqueue`. I don't see the point of changes in this part as the result is still the same: `Command::enqueue`  is called anyway.

Comment (dendibakh) [reply sycl/unittests/scheduler/SchedulerTestUtils.hpp:57]: Well, it's the whole point of mocking: to capture function calls. This is the default for all the tests that call `enqueue` method, i.e. call the original implementation while capturing the fact that the method was called. See more here: https://chromium.googlesource.com/external/github.com/google/googletest/+/HEAD/googlemock/docs/cook_book.md#delegating-calls-to-a-parent-class

**Comment by DenisBakhvalov** [comment]
I need to fix the test failures revealed in the CI.

File: sycl/unittests/scheduler/SchedulerTestUtils.hpp (hunk 3 above)
Comment (s-kanaev) [reply sycl/unittests/scheduler/SchedulerTestUtils.hpp:57]: I understand the point of mocking in general. Though I wonder if you need to mock `Command::enqueue` method at all? Did I miss something in the patch?

Comment (DenisBakhvalov) [reply sycl/unittests/scheduler/SchedulerTestUtils.hpp:57]: This is how I test the changes. See this change for example https://github.com/intel/llvm/pull/2347/commits/794570d51714c7865044d6e55918bbcb64f2ed16. I think it is a good way to write descriptive and small tests.

Comment (s-kanaev) [reply sycl/unittests/scheduler/SchedulerTestUtils.hpp:57]: No objections, then.

**Comment by DenisBakhvalov** [comment]
I fixed CI failures. See https://github.com/intel/llvm/pull/2347/commits/2ec35d1104988f79fd74f1a0b667843f8e6a755d.

File: sycl/unittests/scheduler/BlockedCommands.cpp [hunk 4]
```diff
@@ -45,3 +46,87 @@ TEST_F(SchedulerTest, BlockedCommands) {
...
+  addEdge(&A, &C, nullptr);
+  addEdge(&A, &D, nullptr);
+
+  // We have such a graph:
+  //
+  //     A
+  //   / | \
+  //  B  C  D
+  //
+  // If C is blocked, we should not try to enqueue D.
```
Comment (s-kanaev) [inline sycl/unittests/scheduler/BlockedCommands.cpp:80]: Do I understand it correctly that `A` depends on `B`, `C` and `D`? Why can't we enqueue `D` then?

File: sycl/unittests/scheduler/SchedulerTestUtils.hpp [hunk 5]
```diff
@@ -12,9 +12,10 @@
 #include <detail/queue_impl.hpp>
 #include <detail/scheduler/scheduler.hpp>
 
-#include <functional>
 // This header contains a few common classes/methods used in
 // execution graph testing.
+#include <functional>
+#include <gmock/gmock.h>
```
Comment (s-kanaev) [inline sycl/unittests/scheduler/SchedulerTestUtils.hpp:18]: I think the comment should be below the includes as it describes the `SchedulerTestUtils.hpp` file.

**Review by s-kanaev — COMMENTED** [review]
The changes look good. Only a single comment and a question.

**Review by romanovvlad — COMMENTED** [review]
LGTM

File: sycl/unittests/scheduler/SchedulerTestUtils.hpp (hunk 5 above)
Comment (DenisBakhvalov) [reply sycl/unittests/scheduler/SchedulerTestUtils.hpp:18]: You're right. Will fix.

File: sycl/unittests/scheduler/BlockedCommands.cpp (hunk 4 above)
Comment (DenisBakhvalov) [reply sycl/unittests/scheduler/BlockedCommands.cpp:80]: We discussed it with @romanovvlad and AFAIU, in the current design there are no benefits of including `D`, as it is likely to be blocked as well.

**Comment by DenisBakhvalov** [comment]
@bader, please merge if no objections.

**Review by againull — APPROVED** [review]

**Pull request merged**